src/NUnitFramework/framework/Internal/Execution/ParallelWorkItemDispatcher.cs

```diff
@@ -244,41 +267,45 @@ private void RestoreQueues()
 
         private void OnEndOfShift(object sender, EventArgs ea)
         {
-            // This shift ended, so see if there is work in any other
-            // TODO: Temp fix - revisit later when object model is redesigned.
-            if (StartNextShift())
-                return;
-
-            if (_isolationLevel > 0)
-                RestoreQueues();
+            ShiftFinished?.Invoke(sender as WorkShift);
```

Comment:
Same thing. Can we direct-cast sender to `WorkShift` or at least throw an `ArgumentException`? Would seem to indicate abuse and safe casts create a secondary breach of expectations (raising ShiftFinished with a null WorkShift) while hiding the root cause.